src/zarr/storage/_fsspec.py

```diff
@@ -137,6 +178,39 @@ def from_upath(
             allowed_exceptions=allowed_exceptions,
         )
 
+    @classmethod
+    def from_mapper(
+        cls,
+        fs_map: FSMap,
+        read_only: bool = False,
+        allowed_exceptions: tuple[type[Exception], ...] = ALLOWED_EXCEPTIONS,
+    ) -> FsspecStore:
+        """
+        Create a FsspecStore from a FSMap object.
+
+        Parameters
+        ----------
+        fs_map : FSMap
+            Fsspec mutable mapping object.
+        read_only : bool
+            Whether the store is read-only, defaults to False.
+        allowed_exceptions : tuple, optional
+            The exceptions that are allowed to be raised when accessing the
+            store. Defaults to ALLOWED_EXCEPTIONS.
+
+        Returns
+        -------
+        FsspecStore
+        """
+        if not fs_map.fs.async_impl or not fs_map.fs.asynchronous:
+            fs_map.fs = _make_async(fs_map.fs)
```

Comment:
can we avoid mutating the `fs_map` object here? It looks like we only need the `fs` and `root` attributes, so lets pull those off into their own variables.